wasm/wat/parser.go

```diff
@@ -80,39 +101,101 @@ func (p *moduleParser) ensureLParen(tok tokenType, _ []byte, _, _ uint32) error
+			p.tokenParser = p.parseModule
+		}
 	case fieldModule:
-		p.tokenParser, err = p.moduleFieldHandler(tokenBytes)
+		switch string(fieldName) {
+		// TODO: "types"
+		case "import":
+			p.currentField = fieldModuleImport
+			p.tokenParser = p.parseImport
+		case "start":
+			if p.module.startFunction != nil {
+				return errors.New("redundant start")
+			}
+			p.currentField = fieldModuleStart
+			p.tokenParser = p.parseStart
+		}
 	case fieldModuleImport:
-		p.tokenParser, err = p.importFieldHandler(tokenBytes)
-	default:
-		return fmt.Errorf("unexpected current field %d", p.currentField)
+		// Add the next import func object and ready for parsing it.
+		if string(fieldName) == "func" { // cannot repeat
+			if uint32(len(p.module.importFuncs)) > p.currentImportIndex {
+				return fmt.Errorf("redundant field: %s", string(fieldName))
+			}
+			if err := p.addNextImportFunc(); err != nil {
+				return err
+			}
+			p.currentField = fieldModuleImportFunc
+			p.tokenParser = p.parseImportFunc
+		} // TODO: table, memory or global
+	case fieldModuleImportFunc:
+		switch string(fieldName) {
+		case "param": // can repeat
+			p.currentField = fieldModuleImportFuncParam
+			p.tokenParser = p.parseImportFuncParam
+		case "result": // cannot repeat
+			if p.currentType != nil && len(p.currentType.results) == 1 {
+				return errors.New("redundant result field") // Wasm 1.0 is single or no results
+			}
+			p.currentField = fieldModuleImportFuncResult
+			p.tokenParser = p.parseImportFuncResult
+		case "type": // cannot repeat
+			return errors.New("TODO: (module (import (func (type")
+		}
 	}
-	return
+	if p.tokenParser == nil {
+		return fmt.Errorf("unexpected field: %s", string(fieldName))
+	}
+	return nil
 }
 
-// initialFieldHandler returns a tokenParser for the top-level fields in the WebAssembly source.
-func (p *moduleParser) initialFieldHandler(fieldName []byte) (tokenParser, error) {
-	if string(fieldName) == "module" {
+// endField should be called after encountering tokenRParen. It places the current parser at the parent position based
+// on fixed knowledge of the text format structure.
+//
+// Design Note: This is an alternative to using a stack as the structure parsed by moduleParser is fixed depth. For
+// example, any function body may be parsed in a more dynamic way.
+func (p *moduleParser) endField() error {
+	switch p.currentField {
+	case fieldModuleImportFuncParam, fieldModuleImportFuncResult:
+		p.currentField = fieldModuleImportFunc
+		p.tokenParser = p.parseImportFunc
+	case fieldModuleImportFunc:
+		p.currentField = fieldModuleImport
+		p.tokenParser = p.parseImport
+	case fieldModuleStart, fieldModuleImport:
 		p.currentField = fieldModule
-		return p.parseModule, nil
-	} else {
-		return nil, fmt.Errorf("unexpected field: %s", string(fieldName))
+		p.tokenParser = p.parseModule
+	case fieldModule:
+		p.currentField = fieldInitial
+		p.tokenParser = p.parseUnexpectedToken // only one module is allowed and nothing else
```

Comment:
while there's more to do, at least grouping begin and end allows us to raise this error now